src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fs

```diff
@@ -20,100 +20,143 @@ open Microsoft.Quantum.QsCompiler.Transformations.QsCodeOutput
 
-        /// The right-hand type.
-        Right: ResolvedType
+module RelationOps =
+    let (<.) lhs rhs = Relation(lhs, Subtype, rhs)
 
-        /// The most recent relevant ancestor of the left-hand type.
-        OriginalLeft: ResolvedType
+    let (.=.) lhs rhs = Relation(lhs, Equal, rhs)
 
-        /// The most recent relevant ancestor of the right-hand type.
-        OriginalRight: ResolvedType
+    let (.>) lhs rhs = Relation(lhs, Supertype, rhs)
+
+type TypeMismatch =
+    {
+        Expected: ResolvedType
+        ExpectedContext: ResolvedType option
+        Actual: ResolvedType
+        ActualContext: ResolvedType option
     }
 
-module private TypeContext =
-    /// <summary>
-    /// Creates a type context originating with <paramref name="left"/> and <paramref name="right"/>.
-    /// </summary>
-    let create left right =
+module TypeMismatch =
+    let createWithoutContext expected actual =
         {
-            Left = left
-            Right = right
-            OriginalLeft = left
-            OriginalRight = right
+            Expected = expected
+            ExpectedContext = None
+            Actual = actual
+            ActualContext = None
         }
 
-    /// Descends into the respective children of each type, preserving the original types if the range of both the left
-    /// and the right types do not change.
-    let into (leftChild: ResolvedType) (rightChild: ResolvedType) context =
-        if leftChild.Range = context.Left.Range || rightChild.Range = context.Right.Range then
-            { context with Left = leftChild; Right = rightChild }
-        else
-            create leftChild rightChild
-
-    /// Descends into the respective children of each type, preserving the original types if the range of the right type
-    /// does not change.
-    let intoRight leftChild (rightChild: ResolvedType) context =
-        if rightChild.Range = context.Right.Range then
-            { context with Left = leftChild; Right = rightChild }
-        else
-            create leftChild rightChild
-
-    /// Swaps the left and right types.
-    let swap context =
-        {
-            Left = context.Right
-            Right = context.Left
-            OriginalLeft = context.OriginalRight
-            OriginalRight = context.OriginalLeft
-        }
+    let withContext expected actual mismatch =
+        { mismatch with ExpectedContext = Some expected; ActualContext = Some actual }
+
+type Diagnostic =
+    | TypeMismatch of TypeMismatch
+    | TypeIntersectionMismatch of ordering: Ordering * mismatch: TypeMismatch
+    | CompilerDiagnostic of QsCompilerDiagnostic
+
+module Diagnostic =
+    let withContext expected (actual: ResolvedType) =
+        let sameRange (type_: ResolvedType) : ResolvedType option -> _ =
+            Option.forall (fun context -> type_.Range = context.Range)
+
+        function
+        | TypeMismatch mismatch ->
+            if sameRange actual mismatch.ActualContext then
```

Comment:
I don't understand this line. If the Diagnostic already has a context, then withContext is supposed to replace it, right? Here it seems, if the range of the existing context is the same as the range for the desired context, it gets replaced, or if there isn't an existing context, it gets replaced. But if a context already exists and has a different range, than NoOp?

Reply:
The range checking is designed to find the type of the expression underlined by the diagnostic, and stop updating the parent types after that point. I added a doc comment to `Diagnostic.withParents` about this.